### DISCUSSION TITLE
Question about adding custom log level (between warning and info)

### DISCUSSION THREAD

**Discussion by pmneve** [ElSnoMan/pyleniumio#203] [Q&A] [answered]
I'd like to add a custom log level (at about 24-25) to use for higher level flow tracing than INFO to make debugging a bit easier.

I envision a call like self.py.log.flow(msg) that can be at the beginning of methods for example.

Do you have suggestions/examples?

**Comment by ElSnoMan** [comment]
Take a look at the official `Python Logging` docs here: https://docs.python.org/3/library/logging.html#levels

There are only 6 levels:
![image](https://user-images.githubusercontent.com/23081277/125316135-10b38f00-e2f5-11eb-9241-bc8cc7ae3e70.png)

However, you can create your own loggers to trace the logic in your code.

> 💡 Pylenium has logging for almost every method in its API

```python
import logging

LOGGER = logging.getLogger('framework')
LOGGER.setLevel(logging.DEBUG)

LOGGER.debug('Log a DEBUG message')
LOGGER.info('Log an INFO message')
LOGGER.warning('Log a WARNING message')
LOGGER.error('Log an ERROR message')
LOGGER.critical('Log a CRITICAL message')
LOGGER.fatal('Log a FATAL message')
```

**Reply by pmneve** [reply]
The python docs say that creating a custom log level _is_ possible (with caveats that do not apply in my situation).
![2021-07-12_10-11-11](https://user-images.githubusercontent.com/3780424/125328922-b963ed80-e2f9-11eb-945f-8307ea1ff97c.png)
And there is a method to do so:
![2021-07-12_10-10-20](https://user-images.githubusercontent.com/3780424/125329021-db5d7000-e2f9-11eb-953e-203b817f7a46.png)

There is no other detail in the docs that I can find as to how to use addLevelName() and searching for examples produces several _incomplete_ code snippets that still leave questions.   Was just hoping you had something more straightforward.

I'm not totally clear on your example of creating a separate logger.  I'll explore that direction as it seems simple enought and would give greater control over the logging I'm thinking of.  It looks like you are defining it in a global variable. 

Thanks!

**Reply by ElSnoMan** [reply]
You _can_ use the logger that is attached to Pylenium:

```python
self.py.log.info('Log an INFO message')
```

You can also instantiate your own loggers (however many you want) using the example I posted above with the `logging` package. This provides more granular control.

In Pylenium, we have two loggers:

* One for Pylenium
* One for Report Portal (RP)

That way, if the engineer is using RP as their reporting tool, they automatically get the appropriate logs sent and reported as well.

You might want to log the setup and teardown steps that leverage RESTful APIs. I could see a separate logger for those clients since they are not dependent on Pylenium at all.

For example:

```python
from typing import Dict
import logging
import requests


class UserClient:
    def __init__(self):
        self.log = logging.getLogger('UserClient')

    def get_user_by_email(email: str) -> Dict:
        self.log.info('Create a new user')
        response = requests.get(URL + email)
        if not response.ok:
            self.log.fatal(f'Unable to get user by email: {email}')
        
        return response.json()
```

**Comment by pmneve** [comment]
I’m definitely using the pylenium logger.
The issue is that INFO produces too much information for seeing the flow of the testing code and WARNING none. I want to see the log messages I put in the test methods without the many [STEP] messages (mostly py.get()).   Am trying to find a happy medium which, in CI report, can show the testing flow in every run without being overwhelming. (pytest report is too terse and INFO too verbose).   Am using RP but issue is the same for the stdout/stderr pass through.

It is simpler if it is possible to just have ‘self.py.log.framework’ work as a level between INFO and WARNING.
Using a different logger looks like it might be workable.  I’ll play with that as being nearly as simple and more flexible.

Thanks!
Really appreciate your responsiveness and am very much liking pylenium.
pat


--
Patrick M Neve
Senior QA Automation Engineer
***@***.******@***.***>
Tel. : +1 510 570 3436
Digital Commons, powered by bepress | Elsevier

2100 Milvia Street, Suite 300, Berkeley, CA 94704

From: Carlos Kidman ***@***.***>
Reply-To: ElSnoMan/pyleniumio ***@***.***>
Date: Monday, July 12, 2021 at 10:47
To: ElSnoMan/pyleniumio ***@***.***>
Cc: "Neve, Patrick (ELS-BKY)" ***@***.***>, Author ***@***.***>
Subject: Re: [ElSnoMan/pyleniumio] Question about adding custom log level (between warning and info) (#203)


*** External email: use caution ***



You can use the logger that is attached to Pylenium:

self.py.log.info('Log an INFO message')

You can also instantiate your own loggers (however many you want) using the example I posted above with the logging package. This provides more granular control.

In Pylenium, we have two loggers:
·         One for Pylenium
·         One for Report Portal (RP)

That way, if the engineer is using RP as their reporting tool, they automatically get the appropriate logs sent and reported as well.

You might want to log the setup and teardown steps that leverage RESTful APIs. I could see a separate logger for those clients since they are not dependent on Pylenium at all.

**Comment by ElSnoMan** [comment]
@pmneve Here's a solution I'm thinking of for Pylenium's logging:

1. Change the `py.log.info` messages to `py.log.debug`. It's a lot of detail and information that you most likely don't care about since the exceptions and traces are more helpful. However, if you do want to see the order of operations from the point of view of Pylenium, you can switch to DEBUG
2. Add a schema for logging. This would be simple, but having each log message start with `[PY]:[severity]` would make it easier to search and filter logs from Pylenium or _not_ from Pylenium.

What do you think?

**Comment by pmneve** [comment]
Will take a look at those.  Thanks!

--
Patrick M Neve
Senior QA Automation Engineer
***@***.******@***.***>
Tel. : +1 510 570 3436
Digital Commons, powered by bepress | Elsevier
2100 Milvia Street, Suite 300, Berkeley, CA 94704

From: Carlos Kidman ***@***.***>
Reply-To: ElSnoMan/pyleniumio ***@***.***>
Date: Tuesday, July 13, 2021 at 13:50
To: ElSnoMan/pyleniumio ***@***.***>
Cc: "Neve, Patrick (ELS-BKY)" ***@***.***>, Mention ***@***.***>
Subject: Re: [ElSnoMan/pyleniumio] Question about adding custom log level (between warning and info) (#203)


*** External email: use caution ***



@pmneve<https://nam03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fpmneve&data=04%7C01%7Cp.neve%40elsevier.com%7Ca9b532e5a6c04b17724e08d9463fcd21%7C9274ee3f94254109a27f9fb15c10675d%7C0%7C0%7C637618062116432992%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=9PGlCy3LLJ17sxh%2FFTNhXNREMGgcetNFuwSy4SHmC%2Bc%3D&reserved=0> Here's a solution I'm thinking of for Pylenium's logging:
1.       Change the py.log.info messages to py.log.debug. It's a lot of detail and information that you most likely don't care about since the exceptions and traces are more helpful. However, if you do want to see the order of operations from the point of view of Pylenium, you can switch to DEBUG
2.       Add a schema for logging. This would be simple, but having each log message start with [PY]:[severity] would make it easier to search and filter logs from Pylenium or not from Pylenium.

What do you think?

**Reply by ElSnoMan** [reply]
I mean this solution is something we would change in Pylenium itself and release a new version

**Comment by pmneve** [comment]
Ah.

  1.  would meet my current need nicely!


(2.) probably not necessary as pylenium messages already contain [STEP] and show <element | driver>.py:<line nbr> for identification as coming from pylenium.

Thanks!!

**Comment by pmneve** [comment]
For the moment I've been successful in adding a loglevel between info and warning ("QA", 25) and getting closer to the granularity I need.
I do plan to try adding a custom logger to get more control but this serves for now (gotta deal with those flaky timing issues first)

**Reply by Madhur-Prakash** [reply]
Hey! I came across your post where you added a custom log level, and I’ve been trying to figure out how to do that myself. Could you please share how you implemented it or point me in the right direction? Would really appreciate your help!

**Comment by ElSnoMan** [comment, marked as answer]
I have updated the Logging Documentation so this is more clear 😄 
https://docs.pylenium.io/guides/logging
